trieste/models/normalization/model_wrappers.py

```diff
@@ -0,0 +1,232 @@
+from typing import Mapping, Any, Tuple, Sequence, Union
+import tensorflow as tf
+from gpflow import default_float
+from gpflow.utilities.traversal import read_values, multiple_assign
+from trieste.models.gpflow.interface import GPflowPredictor
+
+from trieste.models.gpflow.models import GaussianProcessRegression
+
+from .transformers import DataTransformer
+from ..interfaces import TrainableProbabilisticModel
+from ...data import Dataset
+from ...types import S, TensorType
+
+
+class DataTransformWrapper(TrainableProbabilisticModel):
+    """
+    A base class that handles data transformations for :class: `TrainableProbabilisticModel`s.
+
+    Usage requires creating a new class that inherits from this class and the desired model, e.g.
+        class NormalizedModel(DataTransformWrapper, GaussianProcessRegression):
+            pass
+    Ensure that this class is the first class inherited from.
+
+    To update model hyperparameters on each iteration, pass `update_parameters=True` into the
+    constructor. It is likely that the user will require a bespoke implementation to handle this
+    with their chosen model. For this, they will typically want to subclass the following methods:
+        _process_hyperparameter_dictionary
+        _get_unnormalized_hyperparameter_priors
+        _update_hyperparameter_priors
+    """
+
+    def __init__(
+        self,
+        dataset: Dataset,
+        *model_args: Any,
+        query_point_transformer: DataTransformer,
```

Comment:
I guess its a realistic case that you might want different type of transformations for inputs and outputs, if we go one step further, even different dimensions?